atintegrators/BeamMomentsPass.c

```diff
@@ -0,0 +1,130 @@
+#include "atelem.c"
+#include "atimplib.c"
+#include <math.h>
+#include <float.h>
+#include <stdio.h>
+#ifdef MPI
+#include <mpi.h>
+#include <mpi4py/mpi4py.h>
+#endif
+/*
+ * Beam moments pass method by Simon White.  
+ *
+ */
+
+struct elem
+{
+  int turn;
+  double *sizes;
+  double *positions;
+};
+
+
+void BeamMomentsPass(double *r_in, int nbunch, int num_particles, struct elem *Elem) {
+
+    int turn = Elem->turn;
+    double *sizes = Elem->sizes;
+    double *positions = Elem->positions;
+        
+    int i, ii, ib; 
+    size_t sz = 2*nbunch*6*sizeof(double) + nbunch*sizeof(double);
+    void *buffer = atMalloc(sz);
+    double *dptr = (double *) buffer;
+    
+    double *avep=dptr; dptr += nbunch*6;
+    double *sizep=dptr; dptr += nbunch*6;
+    double *nparts=dptr;
+    
+    for (i=0; i<nbunch; i++) {
+        nparts[i] = 0.0;
+        for(ii=0; ii<6; ii++) {
+            avep[i*6+ii] = 0.0;
+            sizep[i*6+ii] = 0.0;
+        }
+    }
```

Comment:
You can avoid initialisation by using `atCalloc` which initialises the allocated memory with 0 (as `calloc` does).